nested-data-and-loops/nested-loops.md

```diff
@@ -0,0 +1,203 @@
+
+### Tips
+
+* Use indentation to understand the nesting
+* Take time to understand the local variables in each loop
+* Use words, comments, and pseudocode, and use the phrase "for each ..., we do ..." and "while ... is true, we do ..."
+
+## Visualizing Loops in Loops
+
+It can be a challenge to visualize how many loops run. To visualize nested loops, we can use a loop table. Let's use this code as our example.
+
+``` python
+def map_character_frequency(words):
+    char_map = {}
+    for word in words:
+        for character in word:
+            if character not in char_map.keys():
+                char_map[character] = 1
+            else:
+                char_map[character] += 1
+    return char_map
+
+words = ["deer", "reads"]
+
+char_map = map_character_frequency(words)
+print(char_map)
+```
+
+| `word` | `character` | `char_map` |
+|--------|-------------|------------|
+| `deer` | `d` | `{"d": 1}` |
+| `deer` | `e` | `{"d": 1, "e": 1}` |
+| `deer` | `e` | `{"d": 1, "e": 2}` |
+| `deer` | `r` | `{"d": 1, "e": 2, "r": 1}` |
+| `reads` | `r` | `{"d": 1, "e": 2, "r": 2}` |
+| `reads` | `e` | `{"d": 1, "e": 2, "r": 2}` |
+| `reads` | `a` | `{"d": 1, "e": 2, "r": 2, "a": 1}` |
+| `reads` | `d` | `{"d": 2, "e": 2, "r": 2, "a": 1}` |
+| `reads` | `s` | `{"d": 2, "e": 2, "r": 2, "a": 1, "s": 1}` |
+
+<!-- | Outer loop # | Inner loop # | `word` | `character` | `char_map` |
+|--------------|--------------|--------|-------------|------------|
+| ---          | ---          | ---    | ---         | ---        |
+| 1            | 1            | `deer` | `d` | `{"d": 1}` |
+
+1 | 2 | `deer` | `e` | `{"d": 1, "e": 1}`
+
+1 | 3 | `deer` | `e` | `{"d": 1, "e": 2}`
+
+1 | 4 | `deer` | `r` | `{"d": 1, "e": 2, "r": 1}`
+
+2 | 1 | `reads` | `r` | `{"d": 1, "e": 2, "r": 2}`
+
+2 | 2 | `reads` | `e` | `{"d": 1, "e": 2, "r": 2}`
+
+2 | 3 | `reads` | `a` | `{"d": 1, "e": 2, "r": 2, "a": 1}`
+
+2 | 4 | `reads` | `d` | `{"d": 2, "e": 2, "r": 2, "a": 1}`
+
+2 | 5 | `reads` | `s` | `{"d": 2, "e": 2, "r": 2, "a": 1, "s": 1}` -->
+
+And then at this point, there are no more characters and no more words to loop through.
+
+## Writing Loops in Loops
+
+Writing loops in loops has no special syntax. However, here are some tips and and steps to consider and follow.
+
+Logical errors are the most common errors with nested loops. Attempt to have a clear goal in mind:
+
+* What are you looping through?
+* Why do you need to nest these loops? Could we solve this without nested loops?
+* How can I prove to myself that my loops are working as intended?
+
+### Debugging Loops in Loops
+
+Debugging loops in loops is extremely challenging when we're unclear about how many loops are happening, and the state of the code at any given moment.
+
+When debugging, prioritize the process of stepping through each loop and checking each variable. Check each variable in each loop using:
+
+* a debugger tool
+* paper and pencil, write down each variable
+* a loop table
+* `print` statements
```

Comment:
```suggestion
### Check for Understanding
```

Need review questions below!

Reply:
Prior comment should not remove
* `print` statements

It should be in addition.